python/cudf/cudf/core/df_protocol.py

```diff
@@ -0,0 +1,825 @@
+                 implements the exchange protocol
+
+Notes
+-----
+
+- Interpreting a raw pointer (as in ``Buffer.ptr``) is annoying and unsafe to
+  do in pure Python. It's more general but definitely less friendly than
+  having ``to_arrow`` and ``to_numpy`` methods. So for the buffers which lack
+  ``__dlpack__`` (e.g., because the column dtype isn't supported by DLPack),
+  this is worth looking at again.
+
+"""
+
+
+# A typing protocol could be added later to let Mypy validate code using
+# `from_dataframe` better.
+DataFrameObject = Any
+ColumnObject = Any
+
+
+_INTS = {8: cp.int8, 16: cp.int16, 32: cp.int32, 64: cp.int64}
+_UINTS = {8: cp.uint8, 16: cp.uint16, 32: cp.uint32, 64: cp.uint64}
+_FLOATS = {32: cp.float32, 64: cp.float64}
+_CP_DTYPES = {0: _INTS, 1: _UINTS, 2: _FLOATS, 20: {8: bool}}
+
+
+def from_dataframe(
+    df: DataFrameObject, allow_copy: bool = False
+) -> _CuDFDataFrame:
+    """
+    Construct a cudf DataFrame from ``df`` if it supports ``__dataframe__``
+    """
+    if isinstance(df, cudf.DataFrame):
+        return df
+
+    if not hasattr(df, "__dataframe__"):
+        raise ValueError("`df` does not support __dataframe__")
+
+    return _from_dataframe(df.__dataframe__(allow_copy=allow_copy))
+
+
+def _from_dataframe(df: DataFrameObject) -> _CuDFDataFrame:
+    """
+    Create a cudf DataFrame object from DataFrameObject.
+    """
+    # Check number of chunks, if there's more than one we need to iterate
+    if df.num_chunks() > 1:
+        raise NotImplementedError("More than one chunk not handled yet")
+
+    # We need a dict of columns here, with each column being a cudf column.
+    columns = dict()
+    _k = _DtypeKind
+    _buffers = []  # hold on to buffers, keeps memory alive
+    for name in df.column_names():
+        col = df.get_column_by_name(name)
+
+        if col.dtype[0] in (_k.INT, _k.UINT, _k.FLOAT, _k.BOOL):
+            columns[name], _buf = _protocol_to_cudf_column_numeric(col)
+
+        elif col.dtype[0] == _k.CATEGORICAL:
+            columns[name], _buf = _protocol_to_cudf_column_categorical(col)
+
+        elif col.dtype[0] == _k.STRING:
+            columns[name], _buf = _protocol_to_cudf_column_string(col)
+
+        else:
+            raise NotImplementedError(
+                f"Data type {col.dtype[0]} not handled yet"
+            )
+
+        _buffers.append(_buf)
+
+    df_new = cudf.DataFrame._from_data(columns)
+    df_new._buffers = _buffers
+    return df_new
+
+
+def _protocol_to_cudf_column_numeric(
+    col: _CuDFColumn,
+) -> Tuple[
+    cudf.core.column.ColumnBase,
+    Mapping[str, Optional[Tuple[_CuDFBuffer, ProtoDtype]]],
+]:
+    """
+    Convert an int, uint, float or bool protocol column
+    to the corresponding cudf column
+    """
+    if col.offset != 0:
+        raise NotImplementedError("column.offset > 0 not handled yet")
+
+    buffers = col.get_buffers()
+    assert buffers["data"] is not None, "data buffer should not be None"
```

Comment:
If an `assert` is required for correctness, I suggest replacing it with "if... raise" because [`assert` is removed by Python in optimized mode](https://docs.python.org/3/using/cmdline.html#cmdoption-O). While the "optimized" mode is not commonly used, there is another benefit to this: the `AssertionError` you get from the current pattern is not a very meaningful class name, but something like `ValueError` or `RuntimeError` can carry more meaning. That means you could use something like:
```python
if not condition:
    raise RuntimeError("Message")
```
instead of
```python
assert condition, "Message"
```

Here, I would suggest:
```python
    if buffers["data"] is None:
        raise RuntimeError("data buffer cannot be None")
```

However, I don't think this suggestion is important enough to hold up the PR. It would probably require a lot of changes and I don't think it's necessary to refactor the code for this small consideration. Just wanted to share the idea since it might be helpful in the future.

Reply:
Thank you @bdice for this suggestion. In fact, I was raising exceptions  instead of asserting before. But the thing is that I'm doing it to prevent mypy from complaining. Thus I thought asserting is better as normally this should not happen and they are many checks before to prevent it. 
So, I'll replace assertion by exception.

Reply:
If this is just to make mypy happy, I think the better solution would be to use `typing.cast` to cast away the `Optional`. The underlying issue looks to be that `_CuDFColumn.get_buffers` specifies the return type as `Mapping[str, Optional[Tuple[_CuDFBuffer, ProtoDtype]]]` because both validity and offsets may be None, but data is never none. I don't think there's any way to specify a mapping where only some keys are Optional, so you probably want to do something like `data = cast(Tuple[_CuDFBuffer, ProtoDtype], buffers["data"])` here to make the error go away (preferably with a comment to indicate why).